x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -341,30 +397,64 @@ func (sm *StorageManager) WriteSubscriberPosition(data []byte) error {
 // NewUnlimitedReadWriter returns a read writer with no storage limit.
 // For testing only.
 func (sm *StorageManager) NewUnlimitedReadWriter() StorageLimitReadWriter {
-	return sm.NewReadWriter(0)
+	return sm.NewReadWriter(0, 0)
 }
 
-// NewReadWriter returns a read writer with storage limit.
-func (sm *StorageManager) NewReadWriter(storageLimit uint64) StorageLimitReadWriter {
+// NewReadWriter returns a read writer configured with storage limit and disk threshold ratio.
+func (sm *StorageManager) NewReadWriter(storageLimit uint64, diskThresholdRatio float64) StorageLimitReadWriter {
 	splitRW := SplitReadWriter{
 		eventRW:    sm.eventStorage.NewReadWriter(),
 		decisionRW: sm.decisionStorage.NewReadWriter(),
 	}
 
-	dbStorageLimit := func() uint64 {
-		return storageLimit
-	}
-	if storageLimit == 0 {
-		sm.logger.Infof("setting database storage limit to unlimited")
+	// dbStorageLimit returns max size of db in bytes.
+	// If size of db exceeds dbStorageLimit, writes should be rejected.
+	var dbStorageLimit func() uint64
+	if sm.diskStatFailed.Load() && storageLimit == 0 && diskThresholdRatio > 0 {
```

Comment:
This section feels a bit confusing, what if the disk usage hasn't ran yet and fails after `NewReadWriter` is called?

How about we add the disk usage fallback in the `getDiskUsage` itself and remove the error return? - this way the method has to add fallback by definitione

Reply:
getDiskUsage is called on StorageManager initialization, the first time a TBS processor is created, and before any processor is Run. I can add a comment to make it clear.

Reply:
++ on refactoring in a follow up PR when my head is clearer.

getDiskUsage is designed to mock pebble GetDiskUsage without modifying internal state. `cachedDiskStat`, `getDiskUsageFailed`, `cachedDBSize` are all refreshed regularly throughout the lifetime of StorageManager and should be seen as a (cached) source of truth. Therefore, the use of getDiskUsageFailed here is almost like calling getDiskUsage within NewReadWriter and handling its error.